### PR TITLE
PWX-23390: Raise an event when a pod/VM is not hyperconverged.

### PR DESCRIPTION
**Pull request by adityadani**
**What type of PR is this?**
>improvement

**What this PR does / why we need it**:

- Raise an event on a pod or a VM object when stork knows that the pod/VM needs to be hyperconverged, but is unable to schedule the pod accordingly since Kubernetes did not send the replica nodes as part of the input to stork.
- The event is raised only in prioritize phase.
- The event lists all the nodes where the replicas are present and are potential candidate for hyperconvergence.
- Modified existing UTs to check for an existence of an event.

**Does this PR change a user-facing CRD or CLI?**:
No

**Is a release note needed?**:

```release-note
Improvement: Stork will now raise events on a pod or a VM object if it fails to schedule them in a hyperconverged fashion

```

**Does this change need to be cherry-picked to a release branch?**:
Yes - 24.2.0 & 23.11.1

**Testing Notes**

Following event is raised on a pod
```
Events:
  Type     Reason                Age   From     Message
  ----     ------                ----  ----     -------
  Normal   Scheduled             14s   stork    Successfully assigned ns1/web-0 to ocp-nbhatiacluster3-jndql-worker-zsdjv
  Warning  NonOptimumScheduling  14s   stork    Unable to schedule pod using volumes [pvc-5cb275cd-5249-4386-9481-4231b3deee11] in a hyperconverged fashion.  Make sure you have enough CPU and memory resources available on these nodes: [ocp-nbhatiacluster3-jndql-master-0 ocp-nbhatiacluster3-jndql-worker-w5zkk ocp-nbhatiacluster3-jndql-worker-hkpl9]
```

Following event is raised on a VM
```
Events:
  Type     Reason                Age   From                       Message
  ----     ------                ----  ----                       -------
  Normal   SuccessfulDelete      14m   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance f8495d26-59ee-42f4-a12a-77d0ec1702f6
  Normal   SuccessfulCreate      12m   virtualmachine-controller  Started the virtual machine by creating the new virtual machine instance fio-vm-multi-disk
  Warning  NonOptimumScheduling  12m   stork                      Unable to schedule VM on a node in a hyperconverged fashion. Make sure you have enough CPU and memory resources available on these nodes: [ocp-nbhatiacluster3-jndql-master-1 ocp-nbhatiacluster3-jndql-worker-zsdjv]
```

### CODE REVIEW

**Review by pp511 — APPROVED** [review]
looks good.

File: Makefile [hunk 1]
```diff
@@ -94,6 +94,13 @@ staticcheck:
 			  staticcheck $(PKGS); \
 			  staticcheck -tags integrationtest test/integration_test/*.go;staticcheck -tags unittest $(PKGS)"
 
+mockgen:
```
Comment (diptiranjanpx) [inline Makefile:97]: Does mockgen need to be run as part of `make test`?

**Review by diptiranjanpx — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (adityadani) [reply Makefile:97]: Not really. We just need to run it if we need to update the mock files.

Comment (adityadani) [reply Makefile:97]: I see your point. If the original interfaces change when they get updated in the vendor directory, it is better to generate the mocks again. I will change it in the other PR I have.

**Pull request merged**